src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,811 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file canonical_simplify.cc
+ * \brief Canonical form based simplification.
+ */
+#include <tvm/arithmetic.h>
+#include <tvm/expr_operator.h>
+#include <tvm/ir_mutator.h>
+#include "const_fold.h"
+#include "pattern_match.h"
+#include "rewrite_simplify.h"
+
+namespace tvm {
+namespace arith {
+
+using namespace ir;
+
+class SumExpr;
+class SplitExpr;
+
+/*!
+ * \brief Base class of all temporary expression introduced
+ *        for canonicalization.
+ */
+class CanonicalExprNode : public BaseExprNode {
+ public:
+  /*!
+   * \brief Return the normal Expr that is equivalent to self.
+   * \note Can mutate the internal data structure.
+   * \return The normal expression.
+   */
+  virtual Expr Normalize() const = 0;
+
+  // overrides
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+  }
+  void accept(HalideIR::Internal::IRVisitor* v, const Expr& e) const final {
+    LOG(FATAL) << "not supported";
+  }
+  IRNodeType type_info() const final {
+    return IRNodeType::ExtensionExpr;
+  }
+
+  static constexpr const char* _type_key = "arith.CanonicalExpr";
+  TVM_DECLARE_BASE_NODE_INFO(CanonicalExprNode, BaseExprNode);
+};
+
+/*!
+ * \brief Internal "Split normal form" of expression.
+ *
+ * This is a special expression that represent
+ * a scaled value derived from a split of an index.
+ *
+ * result = ((index % upper_factor) / lower_factor) * scale
+ */
+class SplitExprNode : public CanonicalExprNode {
+ public:
+  /*! \brief The base index expression. */
+  Expr index;
+  /*! \brief The division factor ratio. */
+  int64_t lower_factor{1};
+  /*!
+   * \brief The upper factor.
+   * invariance: (upper_factor == kPosInf || upper_factor % lower_factor == 0)
+   */
+  int64_t upper_factor{kPosInf};
+  /*! \brief scale to the expression. */
+  int64_t scale{1};
+
+  /*! \brief verify that this is a valid entry. */
+  void Verify() const {
+    CHECK(upper_factor == kPosInf || upper_factor % lower_factor == 0);
+  }
+
+  Expr NormalizeWithScale(int64_t sscale) const {
+    Expr res = this->index;
+    Type dtype = this->type;
+    CHECK_EQ(this->type, dtype);
```

Comment:
This check seems redundant.

Reply:
Just like an asset, to check runtime consistency

Reply:
But dtype gets initialized with `this->dtype` in the previous line, so this check is obviously true.